File: Problems/0122.Best-Time-To-Buy-And-Sell-Stock-II/BestTimeToBuyAndSellStockII.go
package leetcode

// Runtime: 4 ms, faster than 95.13% of Go online submissions
func maxProfit(prices []int) int {
	profits := 0
	for i := 1; i < len(prices); i++ {
		if prices[i] > prices[i-1] {
			profits += prices[i] - prices[i-1]
		}
	}
	return profits
}

File: leetcode/0122.Best-Time-To-Buy-And-Sell-Stock-II/BestTimeToBuyAndSellStockII.py
from typing import List

class Solution:

    # Runtime: 56 ms, faster than 96.03% of Python3 online submissions
    def maxProfit(self, prices: List[int]) -> int:
        length = len(prices)
        buy_price = prices[0]
        profits = 0
        idx = 1
        while idx < length:
            curr_price = prices[idx]
            if buy_price < curr_price:
                if idx + 1 < length:
                    next_price = prices[idx + 1]
                else:
                    next_price = 0
                if curr_price > next_price:
                    profits += curr_price - buy_price
                    buy_price = next_price
                    idx += 1
            else:
                buy_price = curr_price
            idx += 1
        return profits

    def maxProfit_v2(self, prices: List[int]) -> int:
        profits = 0
        for idx in range(1, len(prices)):
            if prices[idx] > prices[idx - 1]:
                profits += prices[idx] - prices[idx - 1]
        return profits
